File: singleton02.py
class Aluno:
    _instancia = None
    _inicializado = False

    def __new__(cls, *args, **kwargs):
        if not cls._instancia:
            cls._instancia = super().__new__(cls, *args, **kwargs)
        return cls._instancia
    
    def __init__(self):
        if not self._inicializado:
            self._inicializado = True
            self.nome = ""
            self.notas = {}

    def definirNome(self, nome):
        self.nome = nome

    def addNota(self, disciplina, nota):
        self.notas[disciplina] = nota

    def calcularMedia(self):
        total = sum(self.notas.values())
        return total / len(self.notas)
    
# exemplo de uso
if __name__ == "__main__":
    # instanciando o aluno (utilizando singleton)
    aluno1 = Aluno()
    aluno1.definirNome("Beomgyu")

    # adicionar notas
    aluno1.addNota("Matemática", 8.5)
    aluno1.addNota("História", 7.0)
    aluno1.addNota("Ciências", 9.2)

    # obter a media de notas
    print(f"Media de {aluno1.nome}: {aluno1.calcularMedia()}")

    # criando outra ref ao aluno (mesma instancia utilizada)
    aluno2 = Aluno()
    aluno2.definirNome("Huening Kai")

    # adicionar notas
    aluno2.addNota("Matemática", 9.0)
    aluno2.addNota("História", 5.0)
    aluno2.addNota("Ciências", 8.0)

    # obter a media de notas
    print(f"Media de {aluno2.nome}: {aluno2.calcularMedia()}")

File: singleton02.js
class Aluno{
    constructor(){
        if(!Aluno.instancia){
            this.nome = ""
            this.notas = {};
            Aluno.instancia = this;
        }
        return Aluno.instancia;
    }
    definirNome(nome){
        this.nome = nome;
    }

    addNotas(disciplina, notas){
        this.notas[disciplina] = notas;
    }

    calcularMedia(){
        const total = Object.values(this.notas).reduce((acc, nota) => acc + nota,0);
        return total/ Object.keys(this.notas).length;
    }
}

// exemplo de uso - simulando aplicação
const aluno1 = new Aluno();
aluno1.definirNome("Beomgyu")
aluno1.addNotas("Matemática", 8.5);
aluno1.addNotas("História", 7.0);
aluno1.addNotas("Ciências", 9.2);
console.log(`Media de ${aluno1.nome}: ${aluno1.calcularMedia()}`);

const aluno2 = new Aluno();
aluno1.definirNome("Huening Kai")
aluno1.addNotas("Matemática", 9.0);
aluno1.addNotas("História", 5.0);
aluno1.addNotas("Ciências", 8.0);
console.log(`Media de ${aluno2.nome}: ${aluno2.calcularMedia()}`);

console.log(aluno1 === aluno2);